scripts/functions/cyclegan_transcript.py

```diff
@@ -0,0 +1,343 @@
+    shuffle=True,
+    num_workers=opt.n_cpu,
+)
+# Test data loader
+val_dataloader = DataLoader(
+    TensorDataset(rnaseq_tensor),
+    batch_size=5,
+    shuffle=True,
+    num_workers=1,
+)
+
+
+def sample_images(batches_done):
+    """Saves a generated sample from the test set"""
+    imgs = next(iter(val_dataloader))
+    G_AB.eval()
+    G_BA.eval()
+    real_A = Variable(imgs["A"].type(Tensor))
+    fake_B = G_AB(real_A)
+    real_B = Variable(imgs["B"].type(Tensor))
+    fake_A = G_BA(real_B)
+    # Arange images along x-axis
+    real_A = make_grid(real_A, nrow=5, normalize=True)
+    real_B = make_grid(real_B, nrow=5, normalize=True)
+    fake_A = make_grid(fake_A, nrow=5, normalize=True)
+    fake_B = make_grid(fake_B, nrow=5, normalize=True)
+    # Arange images along y-axis
+    image_grid = torch.cat((real_A, fake_B, real_B, fake_A), 1)
+    save_image(image_grid, "images/%s/%s.png" %
+               (opt.dataset_name, batches_done), normalize=False)
+
+
+# ----------
+#  Training
+# ----------
+
+loss_G_perEpoch = np.zeros([1, opt.n_epochs])
+loss_D_perEpoch = np.zeros([1, opt.n_epochs])
+
+prev_time = time.time()
+
+for epoch in range(opt.epoch, opt.n_epochs):
+    for i, batch in enumerate(dataloader):
+
+        # Set model input
+        real_A = batch[0]  # number of samples in category A
+        real_B = batch[0]
+
+        # Adversarial ground truths
+        valid = Variable(
+            Tensor(np.ones((real_A.size(0), 1))), requires_grad=False)
+        fake = Variable(
+            Tensor(np.zeros((real_A.size(0), 1))), requires_grad=False)
+
+        # ------------------
+        #  Train Generators
+        # ------------------
+
+        G_AB.train()
+        G_BA.train()
+
+        optimizer_G.zero_grad()
+
+        # Identity loss
+        loss_id_A = criterion_identity(G_BA(real_A), real_A)
+        loss_id_B = criterion_identity(G_AB(real_B), real_B)
+
+        loss_identity = (loss_id_A + loss_id_B) / 2
+
+        # GAN loss
+        fake_B = G_AB(real_A)
+        loss_GAN_AB = criterion_GAN(D_B(fake_B), valid)
+        fake_A = G_BA(real_B)
+        loss_GAN_BA = criterion_GAN(D_A(fake_A), valid)
+
+        loss_GAN = (loss_GAN_AB + loss_GAN_BA) / 2
+
+        # Cycle loss
+        recov_A = G_BA(fake_B)
+        loss_cycle_A = criterion_cycle(recov_A, real_A)
+        recov_B = G_AB(fake_A)
+        loss_cycle_B = criterion_cycle(recov_B, real_B)
+
+        loss_cycle = (loss_cycle_A + loss_cycle_B) / 2
+
+        # Total loss
+        loss_G = loss_GAN + opt.lambda_cyc * loss_cycle + opt.lambda_id * loss_identity
+
+        loss_G.backward()
+        optimizer_G.step()
+
+        # -----------------------
+        #  Train Discriminator A
+        # -----------------------
+
+        optimizer_D_A.zero_grad()
+
+        # Real loss
+        loss_real = criterion_GAN(D_A(real_A), valid)
+        # Fake loss (on batch of previously generated samples)
+        fake_A_ = fake_A_buffer.push_and_pop(fake_A)
```

Comment:
Very good point.  This will definitely be key in evaluating if this method is "working" the way we think it should.  Currently I am not.  This PR was just to get cycleGAN running on gene expression data.  Future PR will save the generator model and the transformed data as well.